db/warehouse/migrate/20240506204908_add_form_uniqueness_constraints.rb

```diff
@@ -0,0 +1,19 @@
+class AddFormUniquenessConstraints < ActiveRecord::Migration[7.0]
+  def up
+    safety_assured do
+      remove_index :hmis_form_definitions, name: :uidx_hmis_form_definitions_identifier
+      add_index :hmis_form_definitions, [:identifier, :version], unique: true, where: "deleted_at IS NULL", name: :uidx_hmis_form_definitions_identifier
+      add_index :hmis_form_definitions, [:identifier], unique: true, where: "status = 'draft' AND deleted_at IS NULL", name: :uidx_hmis_form_definitions_one_draft_per_identifier
+      add_index :hmis_form_definitions, [:identifier], unique: true, where: "status = 'published' AND deleted_at IS NULL", name: :uidx_hmis_form_definitions_one_published_per_identifier
```

Comment:
This looks great to me!  Tagging @ttoomey for another pair of eyes. This should represent:
1. `identifier`+`version` are unique together
2. Enforce 0-1 `draft` form per identifier
3. Enforce 0-1 `published` form per identifier